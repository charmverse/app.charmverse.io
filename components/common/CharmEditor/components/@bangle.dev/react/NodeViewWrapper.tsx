import type { NodeView, NodeViewProps } from '@bangle.dev/core';
import React from 'react';

export type RenderNodeViewsFunction = (
  props: NodeViewProps & { children: React.ReactNode },
) => React.ReactNode;

interface PropsType {
  nodeView: NodeView;
  renderNodeViews: RenderNodeViewsFunction;
  nodeViewUpdateStore: WeakMap<NodeView, () => void>;
}

interface StateType {
  nodeViewProps: NodeViewProps;
}

export class NodeViewWrapper extends React.Component<PropsType, StateType> {
  update: () => void;

  attachToContentDOM: (reactElement: HTMLDivElement) => void;

  constructor (props: PropsType) {
    super(props);
    this.update = () => {
      this.setState(() => ({
        nodeViewProps: props.nodeView.getNodeViewProps()
      }));
    };

    this.attachToContentDOM = (reactElement) => {
      if (!reactElement) {
        return;
      }
      const contentDOM = this.props.nodeView.contentDOM!;
      // Since we do not control how many times this callback is called
      // make sure it is not already mounted.
      if (!reactElement.contains(contentDOM)) {
        // If contentDOM happens to be mounted to someone else
        // remove it from there.
        if (contentDOM.parentNode) {
          contentDOM.parentNode.removeChild(contentDOM);
        }
        reactElement.appendChild(contentDOM);
      }
    };
    // So that we can directly update the nodeView without the mess
    // of prop forwarding.
    // What about updating the wrong nodeView ?
    // It is okay because a nodeView and this ReactComponent will always
    // have a 1:1 mapping. This is guaranteed because you use `nodeView` instance
    // to generate a react key. See the usage of this component in ./ReactEditor.js
    props.nodeViewUpdateStore.set(props.nodeView, this.update);
    this.state = { nodeViewProps: this.props.nodeView.getNodeViewProps() };
  }

  getChildren () {
    if (!this.props.nodeView.contentDOM) {
      return null;
    }

    // if (this.state.nodeViewProps.node.isInline) {
    //   return (
    //     // The bangle-nv-content is needed to keep the content take space
    //     // or else browsers will collapse it, making it hard to type
    //     <span
    //       className="bangle-nv-child-container"
    //       ref={this.attachToContentDOM}
    //     />
    //   );
    // }

    return (
      <div
        className='bangle-nv-child-container'
        ref={this.attachToContentDOM}
      />
    );
  }

  render () {
    const element = this.props.renderNodeViews({
      ...this.state.nodeViewProps,
      children: this.getChildren()
    });
    if (!element) {
      throw new Error(
        `renderNodeView must handle rendering for node of type "${this.state.nodeViewProps.node.type.name}"`
      );
    }
    return element;
  }
}
