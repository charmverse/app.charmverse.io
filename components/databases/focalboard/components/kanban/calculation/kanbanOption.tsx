// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React, {useState} from 'react'

import {Option as SelectOption, typesByOptions} from '../../calculations/options'
import {IPropertyTemplate} from '../../../blocks/board'
import ChevronRight from '../../../widgets/icons/chevronRight'
import {Constants} from '../../../constants'

type OptionProps = SelectOption & {
    cardProperties: IPropertyTemplate[]
    onChange: (data: {calculation: string, propertyId: string}) => void
    activeValue: string
    activeProperty: IPropertyTemplate
}

const Option = (props: {data: OptionProps}): JSX.Element => {
    const [submenu, setSubmenu] = useState(false)
    const [height, setHeight] = useState(0)
    const [menuOptionRight, setMenuOptionRight] = useState(0)
    const [calculationToProperties, setCalculationToProperties] = useState<Map<string, IPropertyTemplate[]>>(new Map())

    const toggleOption = (e: any) => {
        if (submenu) {
            setSubmenu(false)
        } else {
            const rect = e.target.getBoundingClientRect()
            setHeight(rect.y)
            setMenuOptionRight(rect.x + rect.width)
            setSubmenu(true)
        }
    }

    if (!calculationToProperties.get(props.data.value)) {
        const supportedPropertyTypes = new Map<string, boolean>([])
        if (typesByOptions.get(props.data.value)) {
            (typesByOptions.get(props.data.value) || []).
                forEach((propertyType) => supportedPropertyTypes.set(propertyType, true))
        }

        const supportedProperties = props.data.cardProperties.
            filter((property) => supportedPropertyTypes.get(property.type) || supportedPropertyTypes.get('common'))

        calculationToProperties.set(props.data.value, supportedProperties)
        setCalculationToProperties(calculationToProperties)
    }

    return (
        <div
            className={`KanbanCalculationOptions_CustomOption ${props.data.activeValue === props.data.value ? 'active' : ''}`}
            onMouseEnter={toggleOption}
            onMouseLeave={toggleOption}
            onClick={() => {
                if (props.data.value !== 'count') {
                    return
                }

                props.data.onChange({
                    calculation: 'count',
                    propertyId: Constants.titleColumnId,
                })
            }}
        >
            <span>
                {props.data.label} {props.data.value !== 'count' && <ChevronRight/>}
            </span>

            {
                submenu && props.data.value !== 'count' && (
                    <div
                        className='dropdown-submenu'
                        style={{top: `${height - 10}px`, left: `${menuOptionRight}px`}}
                    >

                        {
                            calculationToProperties.get(props.data.value) &&
                            calculationToProperties.get(props.data.value)!.map((property) => (
                                <div
                                    key={property.id}
                                    className={`drops ${props.data.activeProperty.id === property.id ? 'active' : ''}`}
                                    onClick={() => {
                                        props.data.onChange({
                                            calculation: props.data.value,
                                            propertyId: property.id,
                                        })
                                    }}
                                >
                                    <span>{property.name}</span>
                                </div>
                            ))
                        }
                    </div>
                )
            }
        </div>
    )
}

export {
    Option,
    OptionProps,
}
