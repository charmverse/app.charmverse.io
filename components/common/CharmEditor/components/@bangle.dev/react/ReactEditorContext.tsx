import type { EditorView } from 'prosemirror-view';
import type { Dispatch, ReactNode, SetStateAction } from 'react';
import { createContext, useContext, useMemo, useState } from 'react';

// keep track of the focused page (may be different from what's in the URL or header)
type ICharmEditorView = {
  view: EditorView | null;
  setView: (view: EditorView) => void;
};

const CharmEditorView = createContext<Readonly<ICharmEditorView>>({
  view: null,
  setView: () => {}
});

export function CharmEditorViewProvider({ children }: { children: ReactNode }) {
  const [view, setView] = useState<EditorView | null>(null);

  const value: ICharmEditorView = useMemo(
    () => ({
      view,
      setView
    }),
    [view, setView]
  );

  return <CharmEditorView.Provider value={value}>{children}</CharmEditorView.Provider>;
}

export const useCharmEditorView = () => useContext(CharmEditorView);
