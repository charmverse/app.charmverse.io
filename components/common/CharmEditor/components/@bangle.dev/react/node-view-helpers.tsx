import { objectUid } from '@bangle.dev/utils';
import { log } from '@charmverse/core/log';
import { useEffect, useState } from 'react';
import type { RefObject } from 'react';
import { flushSync } from 'react-dom';

import type { NodeView, RenderHandlers } from '../core/node-view';
import { saveRenderHandlers } from '../core/node-view';

export const nodeViewUpdateStore = new WeakMap();

type NodeViewsUpdater = (nodeViewUpdateStore: NodeView[]) => NodeView[];
type UpdateNodeViewsFunction = (updater: NodeViewsUpdater) => void;

export const nodeViewRenderHandlers = (updateNodeViews: UpdateNodeViewsFunction): RenderHandlers => ({
  create: (nodeView, _nodeViewProps) => {
    log.debug('create node', objectUid.get(nodeView)); // , new Error().stack);
    updateNodeViews((nodeViews) => [...nodeViews, nodeView]);
  },
  update: (nodeView, _nodeViewProps) => {
    log.debug('update node', objectUid.get(nodeView));
    const updateCallback = nodeViewUpdateStore.get(nodeView);
    // If updateCallback is undefined (which can happen if react took long to mount),
    // we are still okay, as the latest nodeViewProps will be accessed whenever it mounts.
    if (updateCallback) {
      updateCallback();
    }
  },
  destroy: (nodeView) => {
    // log.debug('destroy', objectUid.get(nodeView));
    updateNodeViews((nodeViews) => nodeViews.filter((n) => n !== nodeView));
  }
});

export function useNodeViews(ref: RefObject<HTMLElement>) {
  const [nodeViews, setNodeViews] = useState<NodeView[]>([]);
  useEffect(() => {
    // save the renderHandlers in the dom to decouple nodeView instantiating code
    // from the editor. Since PM passing view when nodeView is created, the author
    // of the component can get the handler reference from `getRenderHandlers(view)`.
    // Note: this assumes that the pm's dom is the direct child of `editorRenderTarget`.
    let destroyed = false;
    saveRenderHandlers(
      ref.current!,
      nodeViewRenderHandlers((cb) => {
        if (!destroyed) {
          // make sure that updates run sequentially for prosemiror, or we get infinite recursion of node views being created and destroyed.
          // See also: https://github.com/ueberdosis/tiptap/pull/2985
          flushSync(() => {
            // use callback variant of setState to
            // always get freshest nodeViews.
            setNodeViews((_nodeViews) => cb(_nodeViews));
          });
        }
      })
    );
    return () => {
      destroyed = true;
    };
  }, [ref]);

  return nodeViews;
}
