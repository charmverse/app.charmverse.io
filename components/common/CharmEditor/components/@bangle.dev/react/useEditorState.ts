import { useState } from 'react';

import type { BangleEditorStateProps } from '../core/bangle-editor-state';
import { BangleEditorState } from '../core/bangle-editor-state';

export function useEditorState(props: BangleEditorStateProps) {
  if (props.plugins && typeof props.plugins !== 'function') {
    throw new Error('plugins error: plugins must be a function');
  }
  const [state] = useState(
    () =>
      // Instantiate the editorState once and keep using that instance
      // on subsequent renders.
      // Passing a callback in useState lazy calls the
      // functions on the first render and never again.
      new BangleEditorState(props)
  );
  return state;
}
