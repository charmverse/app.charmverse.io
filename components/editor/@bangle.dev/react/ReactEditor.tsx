import {
  BangleEditor as CoreBangleEditor,
  BangleEditorProps as CoreBangleEditorProps
} from '@bangle.dev/core';
import { Plugin } from '@bangle.dev/pm';
import { EditorViewContext } from '@bangle.dev/react';
import { nodeViewUpdateStore, useNodeViews } from '@bangle.dev/react/node-view-helpers';
import { NodeViewWrapper, RenderNodeViewsFunction } from '@bangle.dev/react/NodeViewWrapper';
import { objectUid } from '@bangle.dev/utils';
import React, { useEffect, useImperativeHandle, useRef, useState } from 'react';
import reactDOM from 'react-dom';

const LOG = false;

let log = LOG ? console.log.bind(console, 'react-editor') : () => {};

interface BangleEditorProps<PluginMetadata = any>
  extends CoreBangleEditorProps<PluginMetadata> {
  id?: string;
  children?: React.ReactNode;
  renderNodeViews?: RenderNodeViewsFunction;
  className?: string;
  style?: React.CSSProperties;
  onReady?: (editor: CoreBangleEditor<PluginMetadata>) => void;
  editorViewRef?: typeof useRef;
}

export const BangleEditor = React.forwardRef<
  CoreBangleEditor | undefined,
  BangleEditorProps
>(
  (
    {
      id,
      state,
      children,
      focusOnInit = true,
      pmViewOpts,
      renderNodeViews,
      className,
      style,
      onReady = () => {},
    },
    ref,
  ) => {
    const renderRef = useRef<HTMLDivElement>(null);
    const onReadyRef = useRef(onReady);
    const editorViewPayloadRef = useRef({
      state,
      focusOnInit,
      pmViewOpts,
    });
    const [editor, setEditor] = useState<CoreBangleEditor>();
    const nodeViews = useNodeViews(renderRef);

    useImperativeHandle(
      ref,
      () => {
        return editor;
      },
      [editor],
    );

    useEffect(() => {
      const editor = new CoreBangleEditor(
        renderRef.current!,
        editorViewPayloadRef.current,
      );
      (editor.view as any)._updatePluginWatcher = updatePluginWatcher(editor);
      onReadyRef.current(editor);
      setEditor(editor);

      return () => {
        editor.destroy();
      };
    }, [ref]);

    if (nodeViews.length > 0 && renderNodeViews == null) {
      throw new Error(
        'When using nodeViews, you must provide renderNodeViews callback',
      );
    }

    return (
      <React.Fragment>
        {editor ? (
          <EditorViewContext.Provider value={editor.view}>
            {children}
          </EditorViewContext.Provider>
        ) : null}
        <div ref={renderRef} id={id} className={className} style={style} />
        {nodeViews.map((nodeView) => {
          return reactDOM.createPortal(
            <NodeViewWrapper
              debugKey={objectUid.get(nodeView)}
              nodeViewUpdateStore={nodeViewUpdateStore}
              nodeView={nodeView}
              renderNodeViews={renderNodeViews!}
            />,
            nodeView.containerDOM!,
            objectUid.get(nodeView),
          );
        })}
      </React.Fragment>
    );
  },
);

const updatePluginWatcher = (editor: CoreBangleEditor) => {
  return (watcher: Plugin, remove = false) => {
    if (editor.destroyed) {
      return;
    }

    let state = editor.view.state;

    const newPlugins = remove
      ? state.plugins.filter((p) => p !== watcher)
      : [...state.plugins, watcher];

    state = state.reconfigure({
      plugins: newPlugins,
    });

    log('Adding watching to existing state', watcher);
    editor.view.updateState(state);
  };
};
