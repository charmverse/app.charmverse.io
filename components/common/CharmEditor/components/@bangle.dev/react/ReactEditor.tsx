import type {
  BangleEditorProps as CoreBangleEditorProps
} from '@bangle.dev/core';
import {
  BangleEditor as CoreBangleEditor
} from '@bangle.dev/core';
import type { Plugin, Transaction } from '@bangle.dev/pm';
import { EditorViewContext } from '@bangle.dev/react';
import { nodeViewUpdateStore, useNodeViews } from '@bangle.dev/react/node-view-helpers';
import { objectUid } from '@bangle.dev/utils';
import type { ReactNode, RefObject } from 'react';
import React, { useEffect, useImperativeHandle, useRef, useState } from 'react';
import reactDOM from 'react-dom';

import { useUser } from 'hooks/useUser';
import log from 'lib/log';
import { isTouchScreen } from 'lib/utilities/browser';

import { amendTransaction } from '../../suggestions/track/amendTransaction';

import { NodeViewWrapper } from './NodeViewWrapper';
import type { RenderNodeViewsFunction } from './NodeViewWrapper';

interface BangleEditorProps<PluginMetadata = any>
  extends CoreBangleEditorProps<PluginMetadata> {
  id?: string;
  children?: React.ReactNode;
  renderNodeViews?: RenderNodeViewsFunction;
  className?: string;
  style?: React.CSSProperties;
  onReady?: (editor: CoreBangleEditor<PluginMetadata>) => void;
  editorRef?: RefObject<HTMLDivElement>;
  // Components that should be placed underneath the editor
  placeholderComponent?: ReactNode;
  enableSuggestions?: boolean; // requires trackChanges to be true
  trackChanges?: boolean;
}

const updatePluginWatcher = (editor: CoreBangleEditor) => {
  return (watcher: Plugin, remove = false) => {
    if (editor.destroyed) {
      return;
    }

    let state = editor.view.state;

    const newPlugins = remove
      ? state.plugins.filter((p) => p !== watcher)
      : [...state.plugins, watcher];

    state = state.reconfigure({
      plugins: newPlugins
    });

    editor.view.updateState(state);
  };
};

export const BangleEditor = React.forwardRef<
  CoreBangleEditor | undefined,
  BangleEditorProps
>(
  (
    {
      id,
      state,
      children,
      focusOnInit = !isTouchScreen(),
      pmViewOpts,
      renderNodeViews,
      className,
      style,
      onReady = () => {},
      placeholderComponent,
      editorRef,
      enableSuggestions = false,
      trackChanges = false
    },
    ref
  ) => {
    const renderRef = useRef<HTMLDivElement>(null);
    const onReadyRef = useRef(onReady);
    const editorViewPayloadRef = useRef({
      state,
      focusOnInit,
      pmViewOpts,
      enableSuggestions
    });
    const [editor, setEditor] = useState<CoreBangleEditor>();
    const nodeViews = useNodeViews(renderRef);
    const { user } = useUser();

    if (enableSuggestions && !trackChanges) {
      log.error('CharmEditor: Suggestions require trackChanges to be enabled');
    }

    // set current
    editorViewPayloadRef.current.enableSuggestions = enableSuggestions;

    useImperativeHandle(
      ref,
      () => {
        return editor;
      },
      [editor]
    );

    useEffect(() => {
      const _editor = new CoreBangleEditor(
        renderRef.current!,
        editorViewPayloadRef.current
      );
      (_editor.view as any)._updatePluginWatcher = updatePluginWatcher(_editor);
      if (trackChanges) {
        (_editor.view as any)._props.dispatchTransaction = (transaction: Transaction) => {
          const view = _editor.view;
          const trackedUser = { id: user?.id ?? '', username: user?.username ?? '' };
          const trackedTr = amendTransaction(transaction, view.state, trackedUser, editorViewPayloadRef.current.enableSuggestions);
          const { state: newState } = view.state.applyTransaction(trackedTr);
          view.updateState(newState);
        };
      }
      onReadyRef.current(_editor);
      setEditor(_editor);
      return () => {
        _editor.destroy();
      };
    }, [ref]);

    if (nodeViews.length > 0 && renderNodeViews == null) {
      throw new Error(
        'When using nodeViews, you must provide renderNodeViews callback'
      );
    }

    return (
      <EditorViewContext.Provider value={editor?.view as any}>
        <div ref={editorRef} className='bangle-editor-core'>
          {editor ? children : null}
          <div ref={renderRef} id={id} className={className} style={style} />
          {editor ? placeholderComponent : null}
        </div>
        {nodeViews.map((nodeView) => {
          return reactDOM.createPortal(
            <NodeViewWrapper
              nodeViewUpdateStore={nodeViewUpdateStore}
              nodeView={nodeView}
              renderNodeViews={renderNodeViews!}
            />,
              nodeView.containerDOM!,
              objectUid.get(nodeView)
          );
        })}
      </EditorViewContext.Provider>
    );
  }
);
