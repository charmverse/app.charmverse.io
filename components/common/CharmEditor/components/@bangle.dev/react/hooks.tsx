import type { EditorView } from '@bangle.dev/pm';
import { Plugin, PluginKey } from '@bangle.dev/pm';
import { rafSchedule } from '@bangle.dev/utils';
import { useContext, useEffect, useState } from 'react';

import type { BangleEditorStateProps } from 'components/common/CharmEditor/components/@bangle.dev/core/bangle-editor-state';

import { BangleEditorState } from '../core/bangle-editor-state';

import { EditorViewContext } from './ReactEditor';

export function useEditorState(props: BangleEditorStateProps) {
  if (props.plugins && typeof props.plugins !== 'function') {
    throw new Error('plugins error: plugins must be a function');
  }

  const [state] = useState(
    () =>
      // Instantiate the editorState once and keep using that instance
      // on subsequent renders.
      // Passing a callback in useState lazy calls the
      // functions on the first render and never again.
      new BangleEditorState(props)
  );

  return state;
}

export function usePluginState(pluginKey: PluginKey, throttle = false) {
  const view = useEditorViewContext();
  const [state, setState] = useState(pluginKey.getState(view.state));

  useEffect(() => {
    let _setState = setState;
    if (throttle) {
      _setState = rafSchedule(setState);
    }
    const plugin = watcherPlugin(pluginKey, _setState);
    (view as any)._updatePluginWatcher(plugin);
    return () => {
      if (throttle) {
        (_setState as ReturnType<typeof rafSchedule>).cancel();
      }
      (view as any)._updatePluginWatcher(plugin, true);
    };
  }, [view, pluginKey, throttle]);

  return state;
}

export function useEditorViewContext(): EditorView {
  const context = useContext(EditorViewContext);
  if (!context) {
    throw new Error('useEditorViewContext must be used within a EditorViewContext');
  }
  return context;
}

function watcherPlugin(pluginKey: PluginKey, setState: (state: any) => void) {
  return new Plugin({
    key: new PluginKey(`withPluginState_${(pluginKey as any).key}`),
    view() {
      return {
        update(view, prevState) {
          const { state } = view;
          if (prevState === state) {
            return;
          }
          const newPluginState = pluginKey.getState(state);

          if (newPluginState !== pluginKey.getState(prevState)) {
            setState(newPluginState);
          }
        }
      };
    }
  });
}
