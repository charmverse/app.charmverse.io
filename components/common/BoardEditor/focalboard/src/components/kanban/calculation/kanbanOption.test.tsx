// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React from 'react'

import {render} from '@testing-library/react'

import {TestBlockFactory} from '../../../test/testBlockFactory'

import {Option} from './kanbanOption'

describe('components/kanban/calculations/Option', () => {
    const board = TestBlockFactory.createBoard()

    test('base case', () => {
        const component = (
            <Option
                data={{
                    label: 'Count Unique Values',
                    displayName: 'Unique',
                    value: 'countUniqueValue',
                    cardProperties: board.fields.cardProperties,
                    onChange: () => {},
                    activeValue: 'count',
                    activeProperty: board.fields.cardProperties[1],
                }}
            />
        )

        const {container} = render(component)
        expect(container).toMatchSnapshot()
    })
})
