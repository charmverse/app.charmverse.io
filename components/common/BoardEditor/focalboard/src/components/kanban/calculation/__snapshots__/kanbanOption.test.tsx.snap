// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/kanban/calculations/Option base case 1`] = `
<div>
  <div
    class="KanbanCalculationOptions_CustomOption "
  >
    <span>
      Count Unique Values
       
      <i
        class="CompassIcon icon-chevron-right ChevronRightIcon"
      />
    </span>
  </div>
</div>
`;
